File: 0501-1000/0876.Middle of the Linked List.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def middle_node(head)
    def walk(node, count)
        if node.nil?
            puts count / 2
            count / 2
        else
            res = walk(node.next, count + 1)
            (res.is_a? ListNode or count != res) ? res : node
        end
    end
    walk(head, 0)
end

File: 0501-1000/0876.Middle of the Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        p, q = head, head.next
        while q != None:
            p = p.next
            q = q.next and q.next.next
        return p
